lib/os/ring_buffer.c

```diff
@@ -197,11 +205,21 @@ uint32_t ring_buf_put_claim(struct ring_buf *buf, uint8_t **data, uint32_t size)
 
 int ring_buf_put_finish(struct ring_buf *buf, uint32_t size)
 {
+	uint32_t rew;
+	uint32_t threshold = ring_buf_get_rewind_threshold();
+
 	if ((buf->tail + size) > (buf->head + buf->size)) {
 		return -EINVAL;
 	}
 
-	buf->tail += size;
+	/* Check if indexes shall be rewind. */
+	if (buf->tail > threshold) {
+		rew = get_rewind_value(buf->size, threshold);
+	} else {
+		rew = 0;
+	}
+
+	buf->tail += (size - rew);
```

Comment:
Possible improvement.  This calculation may result in `buf->tail` which has a value larger than the rewind threshold which by default is `RING_BUFFER_MAX_SIZE`.  That's not normally an issue, but  with a buffer of size `RING_BUFFER_MAX_SIZE` (unlikely) this condition will result in calculations which overflow `uint32_t` in `ring_buf_is_empty()` and `ring_buf_size_get()`.

```
uint32_t ring_buf_size_get(struct ring_buf *buf)
{
	uint32_t tail = buf->tail;
	uint32_t head = buf->head;

	if (tail < head) {
		tail += get_rewind_value(buf->size,
					 ring_buf_get_rewind_threshold());// get_rewind_value() can return RING_BUFFER_MAX_SIZE and result in tail overflow if buf->size==RING_BUFFER_MAX_SIZE
	}

	return tail - head;
}

```
Suggest only storing values in `buf->head` and `buf->tail` which are < rewind threshold.  The current implementation only considers rewinding `buf->tail` if its already past the rewind threshold _prior_ to adding elements.

Not sure if this is a real issue worth addressing.

Reply:
`ring_buf_init` and declaring macros have asserts to prevent buffers bigger than `RING_BUFFER_MAX_SIZE`

Reply:
Thank you.  Given that `head` will only be rewound after `tail` I can understand now that an overflow should not occur during `tail +=` calculation and this is a non-issue.